### PR TITLE
Optimise and test IP2Country

### PR DESCRIPTION
**Pull request by bemusementpark**
This PR moves `geolite2_country_blocks_ipv4.csv` out of assets and into source, and adds a gradle task `ipToCode` that generates some `geolite2_country_blocks_ipv4.bin` that is placed in generated assets. This file deduplicates IP address ranges, and stores IP addresses and country codes each as ints in binary, which should be close to optimal storage usage.

File size reduction appears to be about 12MB.

This PR removes the `TreeMap` and prefers primitive arrays, as the data is already sorted. This allows us to binary search over the array without additional overhead. Inserting ~100k entries into the TreeMap is an `O ( n log n )` operation, as each insertion is `O ( log n )` so we may as well avoid it. There's also good memory gains, not having an `Entry` a `left`, `right` and `parent` ref for every entry.

All up this improves the performance of this processing from about 3 seconds to about 1 second. The majority of that is from a database lookup in `OnionApi#paths`.

This PR adds a test `IP2CountryTest.kt` that checks some of the corner cases in the ip-country translation.

`ipToCode` task calls `ipToCode.kts` a kotlin script that reads the .csv and writes the .bin

### What's next?

- It'd make sense to inject `IP2Country`, and defer work until you open the `PathActivity`.
- Optimise `OnionApi#paths`

### CODE REVIEW

File: app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt [hunk 1]
```diff
@@ -68,81 +61,79 @@ class IP2Country private constructor(private val context: Context) {
...
+        when {
+            midValue == target -> return mid // Exact match found
+            midValue < target -> low = mid + 1 // Search in the right half
+            else -> high = mid - 1 // Search in the left half
+        }
+    }
+
+    // If no exact match, return the largest index with value <= target
+    return if (high >= 0) high else null
+}
```
Comment (SessionHero01) [inline app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt:139]: Can't we use the Kotlin's `UIntArray.binarySearch`?

**Review by SessionHero01 — COMMENTED** [review]
Awesome work!

**Review by ThomasSession — COMMENTED** [review]
Nice one! Looks like a great optimisation.
Might be worth checking the tests as something seem to have failed there.

File: app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt (hunk 1 above)
Comment (bemusementpark) [reply app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt:139]: `binarySearch` is looking for an exact match or will return `-1`. This is searching for the "Price is Right" match. Best without going over... or the opposite. So we should get an answer for any given IP even if it wasn't explicitly provided in the list.

Comment (SessionHero01) [reply app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt:139]: `binarySearch` doesn't simply return -1 though, it returns "where the item should have been put in order to maintain the order", in other words, it will give you info about the closest matched IP. Refer to the doc of `binarySearch` here:

> the index of the element, if it is contained in the array within the specified range; otherwise, the inverted insertion point (-insertion point - 1). The insertion point is defined as the index at which the element should be inserted, so that the array (or the specified subrange of array) still remains sorted.

**Comment by bemusementpark** [comment]
Thanks @ThomasSession, I moved the `.kts` to a `.gradle.kts` which looks safer, neater, more idiomatic. PTAL.

File: app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt (hunk 1 above)
Comment (bemusementpark) [reply app/src/main/java/org/thoughtcrime/securesms/util/IP2Country.kt]: oh, amazing, thanks @SessionHero01, done.

**Review by SessionHero01 — APPROVED** [review]

**Review by ThomasSession — APPROVED** [review]
